File: treebolicOneOwl/src/main/java/org/treebolic/one/owl/TreebolicBasicActivity.java
/*
 * Copyright (c) 2023. Bernard Bou
 */

package org.treebolic.one.owl;

import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.net.Uri;
import android.os.Build;
import android.os.Bundle;
import android.os.Process;
import android.util.Log;
import android.view.Menu;
import android.view.MenuItem;
import android.view.ViewGroup;
import android.view.inputmethod.InputMethodManager;
import android.webkit.MimeTypeMap;
import android.widget.LinearLayout;
import android.widget.Toast;

import com.google.android.material.snackbar.Snackbar;

import org.treebolic.AppCompatCommonActivity;
import org.treebolic.TreebolicIface;
import org.treebolic.guide.AboutActivity;
import org.treebolic.guide.HelpActivity;
import org.treebolic.guide.Tip;
import org.treebolic.search.ColorUtils;
import org.treebolic.search.SearchSettings;

import java.net.MalformedURLException;
import java.net.URL;
import java.util.Properties;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.ActionBar;
import androidx.appcompat.widget.SearchView;
import androidx.appcompat.widget.Toolbar;
import androidx.core.content.ContextCompat;
import androidx.preference.PreferenceManager;
import treebolic.IContext;
import treebolic.Widget;
import treebolic.glue.component.TreebolicThread;
import treebolic.view.View;

/**
 * Treebolic basic activity
 *
 * @author Bernard Bou
 */
abstract public class TreebolicBasicActivity extends AppCompatCommonActivity implements IContext
{
	/**
	 * Log tag
	 */
	private static final String TAG = "TreebolicBasicA";

	/**
	 * Parameter : Document base
	 */
	@Nullable
	@SuppressWarnings("WeakerAccess")
	protected String base;

	/**
	 * Parameter : Image base
	 */
	@Nullable
	@SuppressWarnings("WeakerAccess")
	protected String imageBase;

	/**
	 * Parameter : Settings
	 */
	@Nullable
	@SuppressWarnings("WeakerAccess")
	protected String settings;

	/**
	 * Parameter : CSS style for WebViews
	 */
	@Nullable
	@SuppressWarnings("WeakerAccess")
	protected String style;

	/**
	 * Parameter : Returned URL urlScheme that is handled
	 */
	@Nullable
	@SuppressWarnings("WeakerAccess")
	protected String urlScheme;

	/**
	 * Parameter : more parameters
	 */
	@Nullable
	@SuppressWarnings("WeakerAccess")
	protected Bundle more;

	/**
	 * Derived parameter : parameters
	 */
	@SuppressWarnings("WeakerAccess")
	protected Properties parameters;

	// components

	/**
	 * Treebolic widget
	 */
	@SuppressWarnings("WeakerAccess")
	protected Widget widget;

	/**
	 * Search view on action bar
	 */
	@SuppressWarnings("WeakerAccess")
	protected SearchView searchView;

	/**
	 * Input
	 */
	@Nullable
	@SuppressWarnings("WeakerAccess")
	protected final String input;

	// parent

	/**
	 * Parent (client) activity
	 */
	@Nullable
	@SuppressWarnings("WeakerAccess")
	protected Intent parentActivity;

	// menu

	/**
	 * Menu id
	 */
	private final int menuId;

	// C O N S T R U C T O R

	protected TreebolicBasicActivity(final int menuId0)
	{
		this.menuId = menuId0;
		this.input = null;
	}

	// L I F E C Y C L E

	@SuppressWarnings("WeakerAccess")
	@Override
	protected void onCreate(final Bundle savedInstanceState)
	{
		super.onCreate(savedInstanceState);

		// widget
		this.widget = new Widget(this, this);

		// content view
		setContentView(R.layout.activity_treebolic);
		final ViewGroup container = findViewById(R.id.container);
		final LinearLayout.LayoutParams params = new LinearLayout.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.MATCH_PARENT, 1.F);
		container.addView(this.widget, params);

		// toolbar
		final Toolbar toolbar = findViewById(R.id.toolbar);

		// action bar
		setSupportActionBar(toolbar);
		final ActionBar actionBar = getSupportActionBar();
		if (actionBar != null)
		{
			actionBar.setElevation(0);
			actionBar.setDisplayOptions(ActionBar.DISPLAY_USE_LOGO | ActionBar.DISPLAY_SHOW_TITLE | ActionBar.DISPLAY_SHOW_HOME | ActionBar.DISPLAY_HOME_AS_UP);
		}
	}

	@Override
	protected void onNewIntent(final Intent intent)
	{
		// an activity will always be paused before receiving a new intent, so you can count on onResume() being called after this method
		super.onNewIntent(intent);

		// getIntent() still returns the original Intent, use setIntent(Intent) to update it to this new Intent.
		setIntent(intent);
	}

	@Override
	protected void onResume()
	{
		Log.d(TreebolicBasicActivity.TAG, "Activity resumed");

		// super
		super.onResume();

		// retrieve arguments
		unmarshalArgs(getIntent());

		// make parameters
		this.parameters = makeParameters();

		// query
		query();

		// first run
		final SharedPreferences prefs = PreferenceManager.getDefaultSharedPreferences(this);
		final boolean hasRun = prefs.getBoolean(Settings.PREF_FIRSTRUN, false);
		if (!hasRun)
		{
			final SharedPreferences.Editor edit = prefs.edit();

			// flag as 'has run'
			edit.putBoolean(Settings.PREF_FIRSTRUN, true).apply();

			// tips
			Tip.show(getSupportFragmentManager());
		}
	}

	@Override
	protected void onPause()
	{
		Log.d(TreebolicBasicActivity.TAG, "Activity paused, terminating surface drawing thread");

		// terminate thread
		final View view = this.widget.getView();
		if (view != null)
		{
			final TreebolicThread thread = view.getThread();
			if (thread != null)
			{
				thread.terminate();
			}
		}

		// super
		super.onPause();
	}

	@Override
	public boolean onCreateOptionsMenu(@NonNull final Menu menu)
	{
		// inflate
		getMenuInflater().inflate(this.menuId, menu);

		// search view
		final MenuItem searchMenuItem = menu.findItem(R.id.action_search);
		searchMenuItem.expandActionView();
		this.searchView = (SearchView) searchMenuItem.getActionView();

		// search view width
		int screenWidth = treebolic.glue.component.Utils.screenWidth(this);
		this.searchView.setMaxWidth(screenWidth / 2);

		// search view listener
		this.searchView.setOnQueryTextListener(new SearchView.OnQueryTextListener()
		{
			@Override
			public boolean onQueryTextSubmit(@NonNull final String query)
			{
				TreebolicBasicActivity.this.searchView.clearFocus();
				TreebolicBasicActivity.this.searchView.setQuery("", false);
				handleQueryChanged(query, true);
				return true;
			}

			@Override
			public boolean onQueryTextChange(@NonNull final String query)
			{
				handleQueryChanged(query, false);
				return true;
			}
		});

		// icon tint
		final int iconTint = ColorUtils.getActionBarForegroundColorFromTheme(this);
		ColorUtils.tint(iconTint, menu, R.id.action_search_run, R.id.action_search_reset, R.id.action_search_settings);

		return super.onCreateOptionsMenu(menu);
	}

	@SuppressWarnings("WeakerAccess")
	@Override
	public boolean onOptionsItemSelected(@NonNull final MenuItem item)
	{
		int itemId = item.getItemId();
		if (itemId == R.id.action_settings)
		{
			final Intent intent = new Intent(this, SettingsActivity.class);
			startActivity(intent);
			return true;
		}
		else if (itemId == R.id.action_finish)
		{
			finish();
			return true;
		}
		else if (itemId == R.id.action_kill)
		{
			Process.killProcess(Process.myPid());
			return true;
		}
		else if (itemId == R.id.action_tips)
		{
			Tip.show(getSupportFragmentManager());
			return true;
		}
		else if (itemId == R.id.action_help)
		{
			startActivity(new Intent(this, HelpActivity.class));
			return true;
		}
		else if (itemId == R.id.action_about)
		{
			startActivity(new Intent(this, AboutActivity.class));
			return true;
		}
		else if (itemId == R.id.action_search_run)
		{
			handleSearchRun();
			return true;
		}
		else if (itemId == R.id.action_search_reset)
		{
			handleSearchReset();
			return true;
		}
		else if (itemId == R.id.action_search_settings)
		{
			SearchSettings.show(getSupportFragmentManager());
			return true;
		}
		return false;
	}

	@Nullable
	@Override
	public Intent getParentActivityIntent()
	{
		if (this.parentActivity != null)
		{
			return this.parentActivity;
		}
		return super.getParentActivityIntent();
	}

	// T R E E B O L I C M O D E L

	/**
	 * Unmarshal model and parameters from intent
	 *
	 * @param intent intent
	 */
	@SuppressWarnings("WeakerAccess")
	protected void unmarshalArgs(@NonNull final Intent intent)
	{
		// retrieve arguments
		final Bundle params = intent.getExtras();
		assert params != null;
		params.setClassLoader(getClassLoader());

		// retrieve arguments
		this.base = params.getString(TreebolicIface.ARG_BASE);
		this.imageBase = params.getString(TreebolicIface.ARG_IMAGEBASE);
		this.settings = params.getString(TreebolicIface.ARG_SETTINGS);
		this.style = params.getString(TreebolicIface.ARG_STYLE);
		this.urlScheme = params.getString(TreebolicIface.ARG_URLSCHEME);
		this.more = Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU ? //
				params.getParcelable(TreebolicIface.ARG_MORE, Bundle.class) : //
				params.getParcelable(TreebolicIface.ARG_MORE);
		this.parentActivity = Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU ? //
				params.getParcelable(TreebolicIface.ARG_PARENTACTIVITY, Intent.class) : //
				params.getParcelable(TreebolicIface.ARG_PARENTACTIVITY);
	}

	// T R E E B O L I C C O N T E X T

	@Nullable
	@Override
	public URL getBase()
	{
		if (this.base != null)
		{
			try
			{
				return new URL(this.base);
			}
			catch (@NonNull final MalformedURLException ignored)
			{
				//
			}
		}
		return Settings.getURLPref(this, TreebolicIface.PREF_BASE);
	}

	@Nullable
	@Override
	public URL getImagesBase()
	{
		if (this.imageBase != null)
		{
			try
			{
				return new URL(this.imageBase);
			}
			catch (@NonNull final MalformedURLException ignored)
			{
				//
			}
		}
		return Settings.getURLPref(this, TreebolicIface.PREF_IMAGEBASE);
	}

	@Override
	public Properties getParameters()
	{
		return this.parameters;
	}

	@Nullable
	@Override
	public String getStyle()
	{
		return this.style != null ? this.style : //
				Settings.STYLE_DEFAULT;
	}

	@Override
	public boolean linkTo(@NonNull final String url, final String target)
	{
		// if url is handled by client, return query to client, which will handle it by initiating another query
		if (this.urlScheme != null && url.startsWith(this.urlScheme))
		{
			final String source2 = url.substring(this.urlScheme.length());
			requery(source2);
			return true;
		}

		// standard handling
		try
		{
			final Intent intent = new Intent(Intent.ACTION_VIEW);
			final Uri uri = Uri.parse(url);
			final String extension = MimeTypeMap.getFileExtensionFromUrl(uri.toString());
			final String mimetype = MimeTypeMap.getSingleton().getMimeTypeFromExtension(extension);
			if (mimetype == null)
			{
				intent.setData(uri);
			}
			else
			{
				intent.setDataAndType(uri, mimetype);
			}
			startActivity(intent);
			return true;
		}
		catch (@NonNull final Exception ignored)
		{
			Toast.makeText(this, R.string.error_link, Toast.LENGTH_LONG).show();
		}
		return false;
	}

	@Nullable
	@Override
	public String getInput()
	{
		return this.input;
	}

	@Override
	public void warn(@NonNull final String message)
	{
		// toast(message, Toast.LENGTH_LONG);
		snackbar(message, Snackbar.LENGTH_LONG);
	}

	@Override
	public void status(@NonNull final String message)
	{
		// toast(message, Toast.LENGTH_SHORT);
		snackbar(message, Snackbar.LENGTH_SHORT);
	}

	// Q U E R Y

	/**
	 * Initial query
	 */
	abstract protected void query();

	/**
	 * Requery (linkTo, or searchView)
	 *
	 * @param source source
	 */
	abstract protected void requery(final String source);

	// S E A R C H

	static private final String CMD_SEARCH = "SEARCH";

	static private final String CMD_RESET = "RESET";

	static private final String CMD_CONTINUE = "CONTINUE";

	//static public int SEARCH_TRIGGER_LEVEL = Integer.MAX_VALUE;

	/**
	 * Search pending flag
	 */
	private boolean searchPending = false;

	/**
	 * SearchView query change listener
	 *
	 * @param query  new query
	 * @param submit whether submit was changed
	 */
	@SuppressWarnings("WeakerAccess")
	protected void handleQueryChanged(@NonNull final String query, boolean submit)
	{
		// clear keyboard out of the way
		if (submit)
		{
			closeKeyboard();
		}

		// reset current search if any
		resetSearch();

		if (submit /*|| query.length() > SEARCH_TRIGGER_LEVEL*/)
		{
			// query applies to source: search is a requery
			final SharedPreferences sharedPrefs = PreferenceManager.getDefaultSharedPreferences(this);
			final String scope = sharedPrefs.getString(SearchSettings.PREF_SEARCH_SCOPE, SearchSettings.SCOPE_LABEL); // label, content, link, id
			if (SearchSettings.SCOPE_SOURCE.equals(scope))
			{
				Log.d(TAG, "Source" + ' ' + '"' + query + '"');
				//if (submit)
				//{
				requery(query);
				//}
				return;
			}

			// query applies to non-source scope (label, content, ..): tree search
			final String mode = sharedPrefs.getString(SearchSettings.PREF_SEARCH_MODE, SearchSettings.MODE_STARTSWITH); // equals, startswith, includes
			runSearch(scope, mode, query);
		}
	}

	/**
	 * Tree search handler
	 */
	@SuppressWarnings("WeakerAccess")
	protected void handleSearchRun()
	{
		// clear keyboard out of the way
		closeKeyboard();

		// new or continued search
		if (!this.searchPending)
		{
			final String query = this.searchView.getQuery().toString();
			final SharedPreferences sharedPrefs = PreferenceManager.getDefaultSharedPreferences(this);
			final String scope = sharedPrefs.getString(SearchSettings.PREF_SEARCH_SCOPE, SearchSettings.SCOPE_LABEL); // label, content, link, id
			if (SearchSettings.SCOPE_SOURCE.equals(scope))
			{
				Log.d(TAG, "Source" + ' ' + '"' + query + '"');
				requery(query);
				return;
			}

			final String mode = sharedPrefs.getString(SearchSettings.PREF_SEARCH_MODE, SearchSettings.MODE_STARTSWITH); // equals, startswith, includes
			runSearch(scope, mode, query);
		}
		else
		{
			continueSearch();
		}
	}

	/**
	 * Tree search reset handler
	 */
	@SuppressWarnings("WeakerAccess")
	protected void handleSearchReset()
	{
		// clear keyboard out of the way
		closeKeyboard();

		// get query
		String query = TreebolicBasicActivity.this.searchView.getQuery().toString();

		// clear current query
		TreebolicBasicActivity.this.searchView.setQuery("", false);

		// query was already empty
		if (query.isEmpty())
		{
			resetSearch();
		}

		// home
		this.widget.focus(null);
	}

	private void closeKeyboard()
	{
		final android.view.View view = this.getCurrentFocus();
		if (view != null)
		{
			final InputMethodManager imm = (InputMethodManager) getSystemService(Context.INPUT_METHOD_SERVICE);
			assert imm != null;
			imm.hideSoftInputFromWindow(view.getWindowToken(), 0);
		}
	}

	// S E A R C H   I N T E R F A C E

	@SuppressWarnings("WeakerAccess")
	protected void runSearch(String scope, String mode, @Nullable String target)
	{
		if (target == null || target.isEmpty())
		{
			return;
		}

		Log.d(TAG, "Search run" + scope + ' ' + mode + ' ' + target);
		this.searchPending = true;
		this.widget.search(CMD_SEARCH, scope, mode, target);
	}

	@SuppressWarnings("WeakerAccess")
	protected void continueSearch()
	{
		Log.d(TAG, "Search continue");
		this.widget.search(CMD_CONTINUE);
	}

	@SuppressWarnings("WeakerAccess")
	protected void resetSearch()
	{
		Log.d(TAG, "Search reset");
		this.searchPending = false;
		this.widget.search(CMD_RESET);
	}

	// H E L P E R S

	/**
	 * Make parameters from bundle
	 *
	 * @return properties
	 */
	@SuppressWarnings("WeakerAccess")
	protected Properties makeParameters()
	{
		final Properties parameters = new Properties();
		if (this.base != null)
		{
			parameters.setProperty("base", this.base);
		}
		if (this.imageBase != null)
		{
			parameters.setProperty("imagebase", this.imageBase);
		}
		if (this.settings != null)
		{
			parameters.setProperty("settings", this.settings);
		}

		// add properties from bundle
		if (this.more != null)
		{
			for (String key : this.more.keySet())
			{
				final String value = this.more.getString(key);
				if (value != null)
				{
					parameters.setProperty(key, value);
				}
			}
		}

		parameters.setProperty("debug", Boolean.toString(BuildConfig.DEBUG));
		return parameters;
	}

	// --Commented out by Inspection START (9/23/17 5:08 PM):
	//	/**
	//	 * Put toast on UI thread
	//	 *
	//	 * @param message  message
	//	 * @param duration duration
	//	 */
	//	private void toast(final String message, final int duration)
	//	{
	//		runOnUiThread(new Runnable()
	//		{
	//			@Override
	//			public void run()
	//			{
	//				Toast.makeText(TreebolicBasicActivity.this, message, duration).show();
	//			}
	//		});
	//	}
	// --Commented out by Inspection STOP (9/23/17 5:08 PM)

	/**
	 * Put snackbar on UI thread
	 *
	 * @param message  message
	 * @param duration duration
	 */
	private void snackbar(@NonNull final String message, final int duration)
	{
		runOnUiThread(() -> {
			Snackbar.make(TreebolicBasicActivity.this.widget, message, duration).show();
			final Snackbar snack = Snackbar.make(TreebolicBasicActivity.this.widget, message, duration);
			final android.view.View view = snack.getView();
			view.setBackgroundColor(ContextCompat.getColor(TreebolicBasicActivity.this, R.color.snackbar_color));
			snack.show();
		});
	}
}

File: treebolicOneSql/src/main/java/org/treebolic/one/sql/TreebolicBasicActivity.kt
/*
 * Copyright (c) 2023. Bernard Bou
 */
package org.treebolic.one.sql

import android.content.Intent
import android.os.Build
import android.os.Bundle
import android.os.Process
import android.util.Log
import android.util.TypedValue
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.view.ViewGroup
import android.view.inputmethod.InputMethodManager
import android.webkit.MimeTypeMap
import android.widget.LinearLayout
import android.widget.Toast
import androidx.appcompat.app.ActionBar
import androidx.appcompat.widget.SearchView
import androidx.appcompat.widget.Toolbar
import androidx.core.content.ContextCompat
import androidx.core.content.edit
import androidx.core.net.toUri
import androidx.preference.PreferenceManager
import com.google.android.material.snackbar.Snackbar
import org.treebolic.AppCompatCommonActivity
import org.treebolic.TreebolicIface
import org.treebolic.guide.AboutActivity
import org.treebolic.guide.HelpActivity
import org.treebolic.guide.Tip
import org.treebolic.one.sql.Settings.getURLPref
import org.treebolic.search.ColorUtils.getActionBarForegroundColorFromTheme
import org.treebolic.search.ColorUtils.tint
import org.treebolic.search.SearchSettings
import treebolic.IContext
import treebolic.Widget
import treebolic.glue.component.Container
import treebolic.glue.component.Surface
import treebolic.glue.component.Utils
import java.net.MalformedURLException
import java.net.URL
import java.util.Properties

/**
 * Treebolic basic activity
 *
 * @property menuId menu id
 *
 * @author Bernard Bou
 */
abstract class TreebolicBasicActivity protected constructor(
    private val menuId: Int
) : AppCompatCommonActivity(), IContext {

    /**
     * Parameter : Document base
     */
    private var base: String? = null

    override fun getBase(): URL? {
        if (base != null) {
            try {
                return URL(base)
            } catch (_: MalformedURLException) {
                
            }
        }
        return getURLPref(this, TreebolicIface.PREF_BASE)
    }

    /**
     * Parameter : Image base
     */
    private var imagesBase: String? = null

    override fun getImagesBase(): URL? {
        if (imagesBase != null) {
            try {
                return URL(imagesBase)
            } catch (_: MalformedURLException) {
                
            }
        }
        return getURLPref(this, TreebolicIface.PREF_IMAGEBASE)
    }

    /**
     * Parameter : Settings
     */
    private var settings: String? = null

    /**
     * Parameter : CSS style for WebViews
     */
    private var cssStyle: String? = null

    override fun getStyle(): String {
        return if (cssStyle != null) cssStyle!! else Settings.STYLE_DEFAULT
    }

    /**
     * Parameter : Returned URL urlScheme that is handled
     */
    private var urlScheme: String? = null

    /**
     * Parameter : more parameters
     */
    private var more: Bundle? = null

    /**
     * Derived parameter : parameters
     */
    private var parameters: Properties? = null

    override fun getParameters(): Properties {
        return parameters!!
    }

    // components

    /**
     * Treebolic widget
     */
    @JvmField
    protected var widget: Widget? = null

    /**
     * Search view on action bar
     */
    protected var searchView: SearchView? = null

    // input

    /**
     * Input
     */
    private val input: String? = null

    override fun getInput(): String? {
        return input
    }

    // parent

    /**
     * Parent (client) activity
     */
    private var parentActivityIntentArg: Intent? = null

    override fun getParentActivityIntent(): Intent? {
        if (parentActivityIntentArg != null) {
            return parentActivityIntentArg
        }
        return super.getParentActivityIntent()
    }

    // L I F E C Y C L E

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        // setup
        val value = TypedValue()
        resources.getValue(R.dimen.splitter_position_percent, value, true)
        val splitterPositionPercent = value.float
        Container.splitterPositionPercent = splitterPositionPercent

        // widget
        widget = Widget(this, this)

        // content view
        setContentView(R.layout.activity_treebolic)
        val container = findViewById<ViewGroup>(R.id.container)
        val params = LinearLayout.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.MATCH_PARENT, 1f)
        val view: View = widget as View
        container.addView(view, params)

        // toolbar
        val toolbar = findViewById<Toolbar>(R.id.toolbar)

        // action bar
        setSupportActionBar(toolbar)
        val actionBar = supportActionBar
        if (actionBar != null) {
            actionBar.elevation = 0f
            actionBar.displayOptions = ActionBar.DISPLAY_USE_LOGO or ActionBar.DISPLAY_SHOW_TITLE or ActionBar.DISPLAY_SHOW_HOME or ActionBar.DISPLAY_HOME_AS_UP
        }
    }

    override fun onNewIntent(intent: Intent) {
        // an activity will always be paused before receiving a new intent, so you can count on onResume() being called after this method
        super.onNewIntent(intent)

        // getIntent() still returns the original Intent, use setIntent(Intent) to update it to this new Intent.
        setIntent(intent)
    }

    override fun onResume() {
        Log.d(TAG, "Activity resumed")

        // super
        super.onResume()

        // retrieve arguments
        unmarshalArgs(intent)

        // make parameters
        parameters = makeParameters()

        // query
        query()

        // first run
        val prefs = PreferenceManager.getDefaultSharedPreferences(this)
        val hasRun = prefs.getBoolean(Settings.PREF_FIRSTRUN, false)
        if (!hasRun) {
            prefs.edit {

                // flag as 'has run'
                putBoolean(Settings.PREF_FIRSTRUN, true)
            }

            // tips
            Tip.show(supportFragmentManager)
        }
    }

    override fun onPause() {
        Log.d(TAG, "Activity paused, terminating surface drawing thread")

        // terminate thread
        val surface: Surface? = widget?.view
        surface?.thread?.terminate()

        // super
        super.onPause()
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        // inflate
        menuInflater.inflate(menuId, menu)

        // search view
        val searchMenuItem = menu.findItem(R.id.action_search)
        searchMenuItem.expandActionView()
        searchView = searchMenuItem.actionView as SearchView?

        // search view width
        val screenWidth = Utils.screenWidth(this)
        searchView!!.maxWidth = screenWidth / 2

        // search view listener
        searchView!!.setOnQueryTextListener(object : SearchView.OnQueryTextListener {
            override fun onQueryTextSubmit(query: String): Boolean {
                searchView!!.clearFocus()
                searchView!!.setQuery("", false)
                handleQueryChanged(query, true)
                return true
            }

            override fun onQueryTextChange(query: String): Boolean {
                handleQueryChanged(query, false)
                return true
            }
        })

        // icon tint
        val iconTint = getActionBarForegroundColorFromTheme(this)
        tint(iconTint, menu, R.id.action_search_run, R.id.action_search_reset, R.id.action_search_settings)

        return super.onCreateOptionsMenu(menu)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            R.id.action_settings -> {
                val intent = Intent(this, SettingsActivity::class.java)
                startActivity(intent)
                return true
            }

            R.id.action_finish -> {
                finish()
                return true
            }

            R.id.action_kill -> {
                Process.killProcess(Process.myPid())
                return true
            }

            R.id.action_tips -> {
                Tip.show(supportFragmentManager)
                return true
            }

            R.id.action_help -> {
                startActivity(Intent(this, HelpActivity::class.java))
                return true
            }

            R.id.action_about -> {
                startActivity(Intent(this, AboutActivity::class.java))
                return true
            }

            R.id.action_search_run -> {
                handleSearchRun()
                return true
            }

            R.id.action_search_reset -> {
                handleSearchReset()
                return true
            }

            R.id.action_search_settings -> {
                SearchSettings.show(supportFragmentManager)
                return true
            }

            else -> return false
        }
    }

    // T R E E B O L I C   M O D E L

    /**
     * Unmarshal model and parameters from intent
     *
     * @param intent intent
     */
    protected open fun unmarshalArgs(intent: Intent) {
        // retrieve arguments
        val params = intent.extras!!
        params.classLoader = classLoader

        // retrieve arguments
        base = params.getString(TreebolicIface.ARG_BASE)
        imagesBase = params.getString(TreebolicIface.ARG_IMAGEBASE)
        settings = params.getString(TreebolicIface.ARG_SETTINGS)
        cssStyle = params.getString(TreebolicIface.ARG_STYLE)
        urlScheme = params.getString(TreebolicIface.ARG_URLSCHEME)
        @Suppress("DEPRECATION")
        parentActivityIntentArg = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU)
            params.getParcelable(TreebolicIface.ARG_PARENTACTIVITY, Intent::class.java) else
            params.getParcelable(TreebolicIface.ARG_PARENTACTIVITY)
        @Suppress("DEPRECATION")
        more = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU)
            params.getParcelable(TreebolicIface.ARG_MORE, Bundle::class.java) else
            params.getParcelable(TreebolicIface.ARG_MORE)
    }

    // T R E E B O L I C   C O N T E X T

    override fun linkTo(url: String, target: String?): Boolean {
        // if url is handled by client, return query to client, which will handle it by initiating another query
        if (urlScheme != null && url.startsWith(urlScheme!!)) {
            val source2 = url.substring(urlScheme!!.length)
            requery(source2)
            return true
        }

        // standard handling
        try {
            val intent = Intent(Intent.ACTION_VIEW)
            val uri = url.toUri()
            val extension = MimeTypeMap.getFileExtensionFromUrl(uri.toString())
            val mimetype = MimeTypeMap.getSingleton().getMimeTypeFromExtension(extension)
            if (mimetype == null) {
                intent.data = uri
            } else {
                intent.setDataAndType(uri, mimetype)
            }
            startActivity(intent)
            return true
        } catch (_: Exception) {
            Toast.makeText(this, R.string.error_link, Toast.LENGTH_LONG).show()
        }
        return false
    }

    override fun warn(message: String) {
        // toast(message, Toast.LENGTH_LONG)
        snackbar(message, Snackbar.LENGTH_LONG)
    }

    override fun status(message: String) {
        // toast(message, Toast.LENGTH_SHORT)
        snackbar(message, Snackbar.LENGTH_SHORT)
    }

    // Q U E R Y

    /**
     * Initial query
     */
    protected abstract fun query()

    /**
     * Requery (linkTo, or searchView)
     *
     * @param newSource new source
     */
    protected abstract fun requery(newSource: String?)

    /**
     * Search pending flag
     */
    private var searchPending = false

    /**
     * SearchView query change listener
     *
     * @param query  new query
     * @param submit whether submit was changed
     */
    protected fun handleQueryChanged(query: String, submit: Boolean) {
        // clear keyboard out of the way
        if (submit) {
            closeKeyboard()
        }

        // reset current search if any
        resetSearch()

        if (submit /*|| query.length() > SEARCH_TRIGGER_LEVEL*/) {
            // query applies to source: search is a requery
            val sharedPrefs = PreferenceManager.getDefaultSharedPreferences(this)
            val scope = sharedPrefs.getString(SearchSettings.PREF_SEARCH_SCOPE, SearchSettings.SCOPE_LABEL) // label, content, link, id
            if (SearchSettings.SCOPE_SOURCE == scope) {
                Log.d(TAG, "Source \"$query\"")
                requery(query)
                return
            }

            // query applies to non-source scope (label, content, ..): tree search
            val mode = sharedPrefs.getString(SearchSettings.PREF_SEARCH_MODE, SearchSettings.MODE_STARTSWITH) // equals, startswith, includes
            runSearch(scope, mode, query)
        }
    }

    /**
     * Tree search handler
     */
    private fun handleSearchRun() {
        // clear keyboard out of the way
        closeKeyboard()

        // new or continued search
        if (!searchPending) {
            val query = searchView!!.query.toString()
            val sharedPrefs = PreferenceManager.getDefaultSharedPreferences(this)
            val scope = sharedPrefs.getString(SearchSettings.PREF_SEARCH_SCOPE, SearchSettings.SCOPE_LABEL) // label, content, link, id
            if (SearchSettings.SCOPE_SOURCE == scope) {
                Log.d(TAG, "Source \"$query\"")
                requery(query)
                return
            }

            val mode = sharedPrefs.getString(SearchSettings.PREF_SEARCH_MODE, SearchSettings.MODE_STARTSWITH) // equals, startswith, includes
            runSearch(scope, mode, query)
        } else {
            continueSearch()
        }
    }

    /**
     * Tree search reset handler
     */
    private fun handleSearchReset() {
        // clear keyboard out of the way
        closeKeyboard()

        // get query
        val query = searchView!!.query.toString()

        // clear current query
        searchView!!.setQuery("", false)

        // query was already empty
        if (query.isEmpty()) {
            resetSearch()
        }

        // home
        widget!!.focus(null)
    }

    private fun closeKeyboard() {
        val view = currentFocus
        if (view != null) {
            val imm = getSystemService(INPUT_METHOD_SERVICE) as InputMethodManager
            imm.hideSoftInputFromWindow(view.windowToken, 0)
        }
    }

    // S E A R C H   I N T E R F A C E

    private fun runSearch(scope: String?, mode: String?, target: String?) {
        if (target.isNullOrEmpty()) {
            return
        }

        Log.d(TAG, "Search run$scope $mode $target")
        searchPending = true
        widget!!.search(CMD_SEARCH, scope, mode, target)
    }

    private fun continueSearch() {
        Log.d(TAG, "Search continue")
        widget!!.search(CMD_CONTINUE)
    }

    private fun resetSearch() {
        Log.d(TAG, "Search reset")
        searchPending = false
        widget!!.search(CMD_RESET)
    }

    // H E L P E R S

    /**
     * Make parameters from bundle
     *
     * @return properties
     */
    protected open fun makeParameters(): Properties? {
        val parameters = Properties()
        if (base != null) {
            parameters.setProperty("base", base)
        }
        if (imagesBase != null) {
            parameters.setProperty("imagebase", imagesBase)
        }
        if (settings != null) {
            parameters.setProperty("settings", settings)
        }

        // add properties from bundle
        if (more != null) {
            for (key in more!!.keySet()) {
                val value = more!!.getString(key)
                if (value != null) {
                    parameters.setProperty(key, value)
                }
            }
        }

        parameters.setProperty("debug", BuildConfig.DEBUG.toString())
        return parameters
    }

    /**
     * Put snackbar on UI thread
     *
     * @param message  message
     * @param duration duration
     */
    private fun snackbar(message: String, duration: Int) {
        runOnUiThread {
            val snack: Snackbar = Snackbar.make(widget as View, message, duration)
            snack.view.setBackgroundColor(ContextCompat.getColor(this@TreebolicBasicActivity, R.color.snackbar_color))
            snack.show()
        }
    }

    companion object {

        private const val TAG = "TreebolicBasicA"

        // S E A R C H

        private const val CMD_SEARCH = "SEARCH"

        private const val CMD_RESET = "RESET"

        private const val CMD_CONTINUE = "CONTINUE"
    }
}
